File: backend/src/test/java/com/team05/petmeeting/domain/feed/controller/FeedControllerTest.java
package com.team05.petmeeting.domain.feed.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.team05.petmeeting.domain.comment.dto.CommentReq;
import com.team05.petmeeting.domain.comment.dto.FeedCommentListRes;
import com.team05.petmeeting.domain.comment.dto.FeedCommentRes;
import com.team05.petmeeting.domain.comment.errorCode.CommentErrorCode;
import com.team05.petmeeting.domain.comment.service.CommentService;
import com.team05.petmeeting.domain.feed.dto.FeedLikeRes;
import com.team05.petmeeting.domain.feed.dto.FeedListRes;
import com.team05.petmeeting.domain.feed.dto.FeedReq;
import com.team05.petmeeting.domain.feed.dto.FeedRes;
import com.team05.petmeeting.domain.feed.enums.FeedCategory;
import com.team05.petmeeting.domain.feed.errorCode.FeedErrorCode;
import com.team05.petmeeting.domain.feed.service.FeedLikeService;
import com.team05.petmeeting.domain.feed.service.FeedService;
import com.team05.petmeeting.domain.user.entity.User;
import com.team05.petmeeting.domain.user.repository.UserRepository;
import com.team05.petmeeting.global.exception.BusinessException;
import com.team05.petmeeting.global.security.handler.JwtAuthenticationEntryPoint;
import com.team05.petmeeting.global.security.test.WithCustomUser;
import com.team05.petmeeting.global.security.userdetails.CustomUserDetails;
import com.team05.petmeeting.global.security.util.JwtUtil;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest;
import org.springframework.data.domain.PageImpl;
import org.springframework.http.MediaType;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.test.context.support.WithAnonymousUser;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

import static org.mockito.Mockito.*;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(FeedController.class)
@AutoConfigureMockMvc(addFilters = false)
@WithCustomUser(userId = 100L)
class FeedControllerTest {

    @Autowired
    private MockMvc mvc;

    private final ObjectMapper objectMapper = new ObjectMapper();

    @MockitoBean
    private UserRepository userRepository;

    @MockitoBean
    private CommentService commentService;

    @MockitoBean
    private FeedService feedService;

    @MockitoBean
    private FeedLikeService feedLikeService;

    @MockitoBean
    private JwtUtil jwtUtil;

    @MockitoBean
    private JwtAuthenticationEntryPoint jwtAuthenticationEntryPoint;

    private Long feedId;
    private Long userId;

    // 인증 객체 만드는 헬퍼 메서드 - 매 테스트마다 반복 작성 안 해도 됨
    private UsernamePasswordAuthenticationToken auth() {
        CustomUserDetails userDetails = new CustomUserDetails(userId, List.of());
        return new UsernamePasswordAuthenticationToken(userDetails, null, List.of());
    }

    @BeforeEach
    void setUp() {
        userId = 100L;
        feedId = 1L;

        User user = User.create("test@test.com", "테스터", "홍길동");
        when(userRepository.findById(userId)).thenReturn(Optional.of(user));
    }

    // ════════════════════════════════════════════════════
    //  피드 작성 테스트
    // ════════════════════════════════════════════════════

    @Test
    @DisplayName("피드 작성 성공")
    void write_success() throws Exception {
        FeedReq req = new FeedReq(FeedCategory.FREE, "새 피드 제목", "새 피드 내용", null, null);

        FeedRes res = new FeedRes(
                null,
                "테스터",
                1L,
                userId,
                null,
                FeedCategory.FREE,
                "새 피드 제목",
                "새 피드 내용",
                null,
                0,
                0,
                List.of(),
                null,
                null
        );
        when(feedService.write(eq(req), any(User.class))).thenReturn(res);

        mvc.perform(post("/api/v1/feeds")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(req))
                        .with(authentication(auth())))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.title").value("새 피드 제목"))
                .andExpect(jsonPath("$.content").value("새 피드 내용"));
    }

    // ════════════════════════════════════════════════════
    //  피드 단건 조회 테스트
    // ════════════════════════════════════════════════════

    @Test
    @DisplayName("피드 단건 조회 성공")
    void getFeed_success() throws Exception {
        FeedRes res = new FeedRes(
                null,
                "테스터",
                feedId,
                userId,
                null,
                FeedCategory.FREE,
                "테스트 피드",
                "내용",
                null,
                0,
                0,
                List.of(),
                null,
                null
        );
        when(feedService.getFeed(feedId)).thenReturn(res);
        mvc.perform(get("/api/v1/feeds/{feedId}", feedId))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.title").value("테스트 피드"))
                .andExpect(jsonPath("$.content").value("내용"));
    }

    @Test
    @DisplayName("피드 단건 조회 실패 - 없는 피드")
    void getFeed_not_found() throws Exception {
        when(feedService.getFeed(999L)).thenThrow(new BusinessException(FeedErrorCode.FEED_NOT_FOUND));
        mvc.perform(get("/api/v1/feeds/{feedId}", 999L))
                .andExpect(status().isNotFound());
    }

    // ════════════════════════════════════════════════════
    //  피드 목록 조회 테스트
    // ════════════════════════════════════════════════════

    @Test
    @DisplayName("피드 목록 조회 성공")
    void getFeeds_success() throws Exception {
        when(feedService.getFeeds(any(), any(), any())).thenReturn(new PageImpl<FeedListRes>(List.of()));
        mvc.perform(get("/api/v1/feeds"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.content").isArray());
    }

    // ════════════════════════════════════════════════════
    //  피드 수정 테스트
    // ════════════════════════════════════════════════════

    @Test
    @DisplayName("피드 수정 성공")
    void modify_success() throws Exception {
        FeedReq req = new FeedReq(FeedCategory.FREE, "수정된 제목", "수정된 내용", null, null);

        FeedRes res = new FeedRes(
                null,
                "테스터",
                feedId,
                userId,
                null,
                FeedCategory.FREE,
                "수정된 제목",
                "수정된 내용",
                null,
                0,
                0,
                List.of(),
                null,
                null
        );
        when(feedService.modify(eq(feedId), eq(req), any(User.class))).thenReturn(res);

        mvc.perform(put("/api/v1/feeds/{feedId}", feedId)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(req))
                        .with(authentication(auth())))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.title").value("수정된 제목"))
                .andExpect(jsonPath("$.content").value("수정된 내용"));
    }

    @Test
    @DisplayName("피드 수정 실패 - 없는 피드")
    void modify_not_found() throws Exception {
        FeedReq req = new FeedReq(FeedCategory.FREE, "수정된 제목", "수정된 내용", null, null);

        when(feedService.modify(eq(999L), eq(req), any(User.class)))
                .thenThrow(new BusinessException(FeedErrorCode.FEED_NOT_FOUND));

        mvc.perform(put("/api/v1/feeds/{feedId}", 999L)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(req))
                        .with(authentication(auth())))
                .andExpect(status().isNotFound());
    }

    // ════════════════════════════════════════════════════
    //  피드 삭제 테스트
    // ════════════════════════════════════════════════════

    @Test
    @DisplayName("피드 삭제 성공")
    void delete_success() throws Exception {
        mvc.perform(delete("/api/v1/feeds/{feedId}", feedId)
                        .with(authentication(auth())))
                .andExpect(status().isNoContent());
    }

    @Test
    @DisplayName("피드 삭제 실패 - 없는 피드")
    void delete_not_found() throws Exception {
        doThrow(new BusinessException(FeedErrorCode.FEED_NOT_FOUND))
                .when(feedService).delete(eq(999L), any(User.class));
        mvc.perform(delete("/api/v1/feeds/{feedId}", 999L)
                        .with(authentication(auth())))
                .andExpect(status().isNotFound());
    }

    // ════════════════════════════════════════════════════
//  추가 실패 케이스
// ════════════════════════════════════════════════════

    @Test
    @DisplayName("피드 수정 실패 - 다른 사람 피드 수정 시도 → 403")
    void modify_forbidden() throws Exception {
        FeedReq req = new FeedReq(FeedCategory.FREE, "수정된 제목", "수정된 내용", null, null);
        when(feedService.modify(eq(feedId), eq(req), any(User.class)))
                .thenThrow(new BusinessException(FeedErrorCode.FORBIDDEN));
        mvc.perform(put("/api/v1/feeds/{feedId}", feedId)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(req))
                        .with(authentication(auth())))
                .andExpect(status().isForbidden()); // 403
    }

    @Test
    @DisplayName("피드 삭제 실패 - 다른 사람 피드 삭제 시도 → 403")
    void delete_forbidden() throws Exception {
        doThrow(new BusinessException(FeedErrorCode.FORBIDDEN))
                .when(feedService).delete(eq(feedId), any(User.class));
        mvc.perform(delete("/api/v1/feeds/{feedId}", feedId)
                        .with(authentication(auth())))
                .andExpect(status().isForbidden()); // 403
    }


    // ════════════════════════════════════════════════════
    //  좋아요 토글 테스트
    // ════════════════════════════════════════════════════

    @Test
    @DisplayName("좋아요 토글 성공")
    void toggleLike_success() throws Exception {
        when(feedLikeService.toggleLike(eq(feedId), any(User.class)))
                .thenReturn(new FeedLikeRes(1, true));
        mvc.perform(post("/api/v1/feeds/{feedId}/likes", feedId)
                        .with(authentication(auth())))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.likeCount").value(1))
                .andExpect(jsonPath("$.isLiked").value(true));
    }


}

File: backend/src/test/java/com/team05/petmeeting/domain/feed/controller/FeedControllerTest.kt
package com.team05.petmeeting.domain.feed.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.team05.petmeeting.domain.comment.service.CommentService
import com.team05.petmeeting.domain.feed.dto.FeedLikeRes
import com.team05.petmeeting.domain.feed.dto.FeedListRes
import com.team05.petmeeting.domain.feed.dto.FeedReq
import com.team05.petmeeting.domain.feed.dto.FeedRes
import com.team05.petmeeting.domain.feed.enums.FeedCategory
import com.team05.petmeeting.domain.feed.errorCode.FeedErrorCode
import com.team05.petmeeting.domain.feed.service.FeedLikeService
import com.team05.petmeeting.domain.feed.service.FeedService
import com.team05.petmeeting.domain.user.entity.User
import com.team05.petmeeting.domain.user.entity.User.Companion.create
import com.team05.petmeeting.domain.user.repository.UserRepository
import com.team05.petmeeting.global.exception.BusinessException
import com.team05.petmeeting.global.security.config.JacksonConfig
import com.team05.petmeeting.global.security.handler.JwtAuthenticationEntryPoint
import com.team05.petmeeting.global.security.test.WithCustomUser
import com.team05.petmeeting.global.security.userdetails.CustomUserDetails
import com.team05.petmeeting.global.security.util.JwtUtil
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.eq
import org.mockito.ArgumentMatchers.isNull
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.http.MediaType
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.handler
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.util.Optional

@WebMvcTest(FeedController::class)
@AutoConfigureMockMvc(addFilters = false)
@Import(JacksonConfig::class)
@WithCustomUser(userId = 100L)
internal class FeedControllerTest {
    @Autowired
    private lateinit var mvc: MockMvc

    @Autowired
    private lateinit var objectMapper: ObjectMapper

    @MockitoBean
    private lateinit var userRepository: UserRepository

    @MockitoBean
    private lateinit var commentService: CommentService

    @MockitoBean
    private lateinit var feedService: FeedService

    @MockitoBean
    private lateinit var feedLikeService: FeedLikeService

    @MockitoBean
    private lateinit var jwtUtil: JwtUtil

    @MockitoBean
    private lateinit var jwtAuthenticationEntryPoint: JwtAuthenticationEntryPoint

    private val userId = 100L
    private val feedId = 1L
    private lateinit var user: User

    @BeforeEach
    fun setUp() {
        user = create("test@test.com", "테스터", "홍길동")
        Mockito.`when`(userRepository.findById(userId)).thenReturn(Optional.of(user))
    }

    @Test
    @DisplayName("피드 작성 성공")
    fun write_success() {
        val req = FeedReq(FeedCategory.FREE, "새 피드 제목", "새 피드 내용", null, null)
        val res = feedRes(
            title = "새 피드 제목",
            content = "새 피드 내용"
        )

        Mockito.`when`(feedService.write(req, user)).thenReturn(res)

        mvc.perform(
            post("/api/v1/feeds")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(req))
                .with(authentication(auth()))
        )
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("write"))
            .andExpect(status().isCreated)
            .andExpect(jsonPath("$.title").value("새 피드 제목"))
            .andExpect(jsonPath("$.content").value("새 피드 내용"))
    }

    @Test
    @DisplayName("피드 단건 조회 성공")
    fun getFeed_success() {
        Mockito.`when`(feedService.getFeed(feedId)).thenReturn(
            feedRes(title = "테스트 피드", content = "내용")
        )

        mvc.perform(get("/api/v1/feeds/{feedId}", feedId))
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("getFeed"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.title").value("테스트 피드"))
            .andExpect(jsonPath("$.content").value("내용"))
    }

    @Test
    @DisplayName("피드 단건 조회 실패 - 없는 피드")
    fun getFeed_not_found() {
        Mockito.`when`(feedService.getFeed(999L))
            .thenThrow(BusinessException(FeedErrorCode.FEED_NOT_FOUND))

        mvc.perform(get("/api/v1/feeds/{feedId}", 999L))
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("getFeed"))
            .andExpect(status().isNotFound)
    }

    @Test
    @DisplayName("피드 목록 조회 성공")
    fun getFeeds_success() {
        Mockito.`when`(
            feedService.getFeeds(
                anyPageable(),
                eq(userId),
                isNull<FeedCategory>()
            )
        ).thenReturn(PageImpl<FeedListRes>(emptyList()))


        mvc.perform(get("/api/v1/feeds"))
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("getFeeds"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.content").isArray)
    }

    @Test
    @DisplayName("피드 수정 성공")
    fun modify_success() {
        val req = FeedReq(FeedCategory.FREE, "수정된 제목", "수정된 내용", null, null)
        val res = feedRes(title = "수정된 제목", content = "수정된 내용")

        Mockito.`when`(feedService.modify(feedId, req, user)).thenReturn(res)

        mvc.perform(
            put("/api/v1/feeds/{feedId}", feedId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(req))
                .with(authentication(auth()))
        )
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("modify"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.title").value("수정된 제목"))
            .andExpect(jsonPath("$.content").value("수정된 내용"))
    }

    @Test
    @DisplayName("피드 수정 실패 - 없는 피드")
    fun modify_not_found() {
        val req = FeedReq(FeedCategory.FREE, "수정된 제목", "수정된 내용", null, null)

        Mockito.`when`(feedService.modify(999L, req, user))
            .thenThrow(BusinessException(FeedErrorCode.FEED_NOT_FOUND))

        mvc.perform(
            put("/api/v1/feeds/{feedId}", 999L)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(req))
                .with(authentication(auth()))
        )
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("modify"))
            .andExpect(status().isNotFound)
    }

    @Test
    @DisplayName("피드 삭제 성공")
    fun delete_success() {
        mvc.perform(
            delete("/api/v1/feeds/{feedId}", feedId)
                .with(authentication(auth()))
        )
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("delete"))
            .andExpect(status().isNoContent)
    }

    @Test
    @DisplayName("피드 삭제 실패 - 없는 피드")
    fun delete_not_found() {
        Mockito.doThrow(BusinessException(FeedErrorCode.FEED_NOT_FOUND))
            .`when`(feedService)
            .delete(999L, user)

        mvc.perform(
            delete("/api/v1/feeds/{feedId}", 999L)
                .with(authentication(auth()))
        )
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("delete"))
            .andExpect(status().isNotFound)
    }

    @Test
    @DisplayName("피드 수정 실패 - 다른 사람 피드 수정 시도 → 403")
    fun modify_forbidden() {
        val req = FeedReq(FeedCategory.FREE, "수정된 제목", "수정된 내용", null, null)

        Mockito.`when`(feedService.modify(feedId, req, user))
            .thenThrow(BusinessException(FeedErrorCode.FORBIDDEN))

        mvc.perform(
            put("/api/v1/feeds/{feedId}", feedId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(req))
                .with(authentication(auth()))
        )
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("modify"))
            .andExpect(status().isForbidden)
    }

    @Test
    @DisplayName("피드 삭제 실패 - 다른 사람 피드 삭제 시도 → 403")
    fun delete_forbidden() {
        Mockito.doThrow(BusinessException(FeedErrorCode.FORBIDDEN))
            .`when`(feedService)
            .delete(feedId, user)

        mvc.perform(
            delete("/api/v1/feeds/{feedId}", feedId)
                .with(authentication(auth()))
        )
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("delete"))
            .andExpect(status().isForbidden)
    }

    @Test
    @DisplayName("좋아요 토글 성공")
    fun toggleLike_success() {
        Mockito.`when`(feedLikeService.toggleLike(feedId, user))
            .thenReturn(FeedLikeRes(1, true))

        mvc.perform(
            post("/api/v1/feeds/{feedId}/likes", feedId)
                .with(authentication(auth()))
        )
            .andExpect(handler().handlerType(FeedController::class.java))
            .andExpect(handler().methodName("toggleLike"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.likeCount").value(1))
            .andExpect(jsonPath("$.isLiked").value(true))
    }

    private fun auth(): UsernamePasswordAuthenticationToken {
        val userDetails = CustomUserDetails(userId, emptyList<GrantedAuthority>())
        return UsernamePasswordAuthenticationToken(userDetails, null, emptyList<GrantedAuthority>())
    }

    private fun anyPageable(): Pageable {
        return any(Pageable::class.java) ?: Pageable.unpaged()
    }

    private fun feedRes(
        title: String,
        content: String,
        category: FeedCategory = FeedCategory.FREE
    ): FeedRes {
        return FeedRes(
            profileImageUrl = null,
            nickname = "테스터",
            feedId = feedId,
            userId = userId,
            animalId = null,
            category = category,
            title = title,
            content = content,
            imageUrl = null,
            likeCount = 0,
            commentCount = 0,
            comments = emptyList(),
            createdAt = null,
            updatedAt = null
        )
    }
}
